### ISSUE TITLE
Support mulitple port definitions on a Service

### ISSUE THREAD

**Issue by fgiorgetti** [skupperproject/skupper#569]
Allow adding multiple ports to a Skupper service.

**Comment by ted-ross** [comment]
Hi @fgiorgetti - I'm interested in multi-port services as well.  Do you have a plan for how to map the different ports to VAN addresses?  I'd be happy to discuss if it's helpful.

-Ted

**Comment by fgiorgetti** [comment]
Hey @ted-ross I am starting to write down a proposal for accommodating multiple
ports into a single service, without breaking compatibility with older Skupper sites.
The idea is that an older Skupper site will ignore services with multiple ports.

It would be great to discuss and hear your ideas. I believe @ajssmith can arrange 
or invite you to a Skupper team meeting session and add this topic into the agenda.

**Comment by ted-ross** [comment]
I suspect that this may need to be a non-compatible change.  Multi-port services expose multi-port pods running on other clusters.  Both sides of the interaction are going to have to change to handle this capability.

**Comment by fgiorgetti** [comment]
Correct, if you want it to work on all your sites, then you will need to have all of them updated,
otherwise your outdated site won't be compatible and will not provide access to all the ports.

**Comment by ted-ross** [comment]
While we're considering changes to the VAN-address mappings to support multi-ported pods, we should also consider the changes needed to target multiple different label-sets (selectors) from a service (blue/green, canary, A/B, etc.).

If we're going to make an incompatible change to the address mappings, we should cover as many use cases as possible in the change.

**Comment by grs** [comment]
We can already support different selectors on the same service (via different targets). What we don't have is some way to weight traffic differently to different targets.

**Comment by ted-ross** [comment]
Questions on the updated description:

1. Why are there two attributes for port(s) and targetPort(s), singular and multiple?  Does the port need to be among those listed in ports?  Which of the ports should be placed in port?
2. For the VAN addresses, why not use the common ':' delimiter to delimit the port instead of '@' (i.e. use "my-app:8080")

Also, I would suggest that you use different service and pod ports in your example to make it easier to ascertain which port is being referred to.

**Comment by grs** [comment]
> 1. Why are there two attributes for port(s) and targetPort(s), singular and multiple?  Does the port need to be among those listed in ports?  Which of the ports should be placed in port?

For backward compatibility or at least the potential for detection and more graceful handling of older sites in a network.

**Comment by fgiorgetti** [comment]
> 1. Why are there two attributes for port(s) and targetPort(s), singular and multiple?  Does the port need to be among those listed in ports?  Which of the ports should be placed in port?

The idea was to keep service sync message valid from the standpoint of an older site (as @grs said),
but it definitely causes the content to be confusing and redundant. I will evaluate what Gordon suggested
and prepare an updated proposal for discussion.

> 2. For the VAN addresses, why not use the common ':' delimiter to delimit the port instead of '@' (i.e. use "my-app:8080")

Agreed. Makes more sense to use ':' in the address.
What about the listener and connector names? I see today they are named using '@' to separate service/target information
from the target hostname. In my proposal, I am keeping the '@' and adding ':<port>' to it, to avoid name collisions
for targets going to same host on different ports.

Any recommendation?

> Also, I would suggest that you use different service and pod ports in your example to make it easier to ascertain which port is being referred to.

I will update it so that this information is clearer.

**Comment by grs** [comment]
> What about the listener and connector names? I see today they are named using '@' to separate service/target information
> from the target hostname.

For connectors, the target will be host:port (e.g. foo@bar:9090). For listeners the host is omitted as it always 0.0.0.0 (so thats where you have e.g. baz@9191). I don't think there is a need to change the naming strategy (at least not as part of this).

**Comment by ted-ross** [comment]
> For connectors, the target will be host:port (e.g. foo@bar:9090). For listeners the host is omitted as it always 0.0.0.0 (so thats where you have e.g. baz@9191). I don't think there is a need to change the naming strategy (at least not as part of this).

Agreed.  The connector/listener naming is really a local concern.  The names only need to be unique and derivable from the data structure.

**Comment by fgiorgetti** [comment]
@grs and @ted-ross I have updated the description with the recommended changes.
Please review and share your thoughts and feedback.

**Comment by grs** [comment]
In the example router config, it looks like it has only configured connectors for the explicitly mapped target ports. All ports exposed by the service need to be handled, so the v1 app will be expected to server port 9090 (as it did not map that) and the v2 app will be expected to serve port 8080.

Other than that, this looks good to me.

**Comment by fgiorgetti** [comment]
> In the example router config, it looks like it has only configured connectors for the explicitly mapped target ports. All ports exposed by the service need to be handled, so the v1 app will be expected to server port 9090 (as it did not map that) and the v2 app will be expected to serve port 8080.

I have considered the following scenario for this sample:

```
skupper service create my-app 8080 9090
skupper service bind my-app deployment my-app-v1 --target-port 8080:8181
skupper service bind my-app deployment my-app-v2 --target-port 9090:9191
```

I agree with you in the following scenario:

```
skupper service create my-app 8080 9090
skupper service bind my-app deployment my-app-v1
skupper service bind my-app deployment my-app-v2
```

In this case both ports will be mapped to both targets.

What do you think?

**Comment by grs** [comment]
I think that the all ports need to be implemented by all targets. They can map a particular port to a different targetPort, but they can't opt out of providing a port at all.

So even in the first instance I would expect that both v1 and v2 have two ports exposed, 8181&9090 in the case of v1 and 8080&9191 in the case of v2.

**Comment by grs** [comment]
I guess I *could* envisage some way of saying that a target did not implement a given port, but I think that  would be a more esoteric use case and an explicit declaration would be better. E.g. --target-port 9090:-1

**Comment by grs** [comment]
Also just for completeness, I think the primary use case is going to be exposing a multi-port deployment or statefulset, e.g.

skupper expose deployment my-app-v1

Where the ports will often be deduced from the container ports defined in that deployment.

**Comment by fgiorgetti** [comment]
@grs I agree. Although it is _possible_ to do that, it will make things more complex and
this is not a trivial use case for sure. I will update it again.

**Comment by fgiorgetti** [comment]
Updated again. I have also added a note in the `service expose` about the changes to its flags.

**Comment by grs** [comment]
LGTM

**Comment by fgiorgetti** [comment]
PR #585 is ready for review now. One difference with regards to this original proposal
is that old sites won't be compatible for service synchronization due to the new addressing
strategy which qualifies the address with a respective service port.

I am going to propose a separate enhancement to help preventing connections from
incompatible skupper versions.
